boa/src/object/internal_methods/arguments.rs

```diff
@@ -0,0 +1,261 @@
+            PropertyDescriptor::builder()
+                .value(map.get(key.clone(), context)?)
+                .maybe_writable(desc.writable())
+                .maybe_enumerable(desc.enumerable())
+                .maybe_configurable(desc.configurable())
+                .build()
+        } else {
+            // 6. Return desc.
+            desc
+        },
+    ))
+}
+
+/// [[DefineOwnProperty]] for arguments exotic objects.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-arguments-exotic-objects-defineownproperty-p-desc
+pub(crate) fn arguments_exotic_define_own_property(
+    obj: &JsObject,
+    key: PropertyKey,
+    desc: PropertyDescriptor,
+    context: &mut Context,
+) -> JsResult<bool> {
+    // 1. Let map be args.[[ParameterMap]].
+    let map = obj
+        .borrow()
+        .as_mapped_arguments()
+        .expect("arguments exotic method must only be callable from arguments objects")
+        .parameter_map();
+
+    // 2. Let isMapped be HasOwnProperty(map, P).
+    let is_mapped = map.has_own_property(key.clone(), context)?;
+
+    let new_arg_desc = match desc.kind() {
+        // 4. If isMapped is true and IsDataDescriptor(Desc) is true, then
+        // a. If Desc.[[Value]] is not present and Desc.[[Writable]] is present and its
+        // value is false, then
+        DescriptorKind::Data {
+            writable: Some(false),
+            value: None,
+        } if is_mapped =>
+        // i. Set newArgDesc to a copy of Desc.
+        // ii. Set newArgDesc.[[Value]] to Get(map, P).
+        {
+            PropertyDescriptor::builder()
+                .value(map.get(key.clone(), context)?)
+                .writable(false)
+                .maybe_enumerable(desc.enumerable())
+                .maybe_configurable(desc.configurable())
+                .build()
+        }
+
+        // 3. Let newArgDesc be Desc.
+        _ => desc.clone(),
+    };
+
+    // 5. Let allowed be ? OrdinaryDefineOwnProperty(args, P, newArgDesc).
+    // 6. If allowed is false, return false.
+    if !super::ordinary_define_own_property(obj, key.clone(), new_arg_desc, context)? {
+        return Ok(false);
+    }
+
+    // 7. If isMapped is true, then
+    if is_mapped {
+        // a. If IsAccessorDescriptor(Desc) is true, then
+        if desc.is_accessor_descriptor() {
+            // i. Call map.[[Delete]](P).
+            map.__delete__(&key, context)?;
+
+        // b. Else,
+        } else {
+            // i. If Desc.[[Value]] is present, then
+            if let Some(value) = desc.value() {
+                // 1. Let setStatus be Set(map, P, Desc.[[Value]], false).
+                let set_status = map.set(key.clone(), value, false, context)?;
+
+                // 2. Assert: setStatus is true because formal parameters mapped by argument
+                // objects are always writable.
+                debug_assert!(set_status)
```

Comment:
The spec does not expect `set` to fail here. So we should either `expect` on `set` or assert that `set_status` is `Ok(true)`.